include/nbl/builtin/hlsl/workgroup/shuffle.hlsl

```diff
@@ -18,8 +18,29 @@ namespace workgroup
 {
 
 
-template<typename SharedMemoryAccessor, typename Scalar, uint32_t N = 1>
-void shuffleXor(NBL_REF_ARG(vector <Scalar, N>) value, uint32_t mask, uint32_t threadID, NBL_REF_ARG(MemoryAdaptor<SharedMemoryAccessor>) sharedmemAdaptor)
+template<typename SharedMemoryAccessor, typename T>
+struct shuffleXor
+{
+    static void __call(NBL_REF_ARG(T) value, uint32_t mask, uint32_t threadID, NBL_REF_ARG(MemoryAdaptor<SharedMemoryAccessor>) sharedmemAdaptor)
+    {
+        sharedmemAdaptor.set(threadID, value);
+        
+        // Wait until all writes are done before reading
+        sharedmemAdaptor.workgroupExecutionAndMemoryBarrier();
+    
+        sharedmemAdaptor.get(threadID ^ mask, value);
+    }
+
+    static void __call(NBL_REF_ARG(T) value, uint32_t mask, NBL_REF_ARG(MemoryAdaptor<SharedMemoryAccessor>) sharedmemAdaptor)
```

Comment:
you're not supposed to require that the type being passed in is a `MemoryAdaptor<AccessorT>`, keep it as `AccessorT` always